File: dart/lib/snapkit_image_url.dart
/// Snapkit Image URL Builder for Dart/Flutter
library snapkit_image_url;

/// Resize mode
enum Fit {
  contain,
  cover,
  fill,
  inside,
  outside,
}

/// Output format
enum Format {
  jpeg,
  png,
  webp,
  avif,
}

/// Region extraction
class Extract {
  final int x;
  final int y;
  final int width;
  final int height;

  const Extract({
    required this.x,
    required this.y,
    required this.width,
    required this.height,
  });
}

/// Image transformation parameters
class TransformOptions {
  /// Image width (pixels)
  final int? w;

  /// Image height (pixels)
  final int? h;

  /// Resize mode
  final Fit? fit;

  /// Output format
  final Format? format;

  /// Rotation angle (degrees)
  final int? rotation;

  /// Blur strength (0.3-1000)
  final int? blur;

  /// Grayscale conversion flag
  final bool? grayscale;

  /// Vertical flip flag
  final bool? flip;

  /// Horizontal flip flag
  final bool? flop;

  /// Region extraction
  final Extract? extract;

  /// Device Pixel Ratio (1.0-4.0)
  final double? dpr;

  /// Image quality (1-100)
  final int? quality;

  const TransformOptions({
    this.w,
    this.h,
    this.fit,
    this.format,
    this.rotation,
    this.blur,
    this.grayscale,
    this.flip,
    this.flop,
    this.extract,
    this.dpr,
    this.quality,
  });
}

/// Snapkit image URL builder
///
/// Example:
/// ```dart
/// final builder = SnapkitImageURL('my-org');
/// final imageUrl = builder.build(
///   url: 'https://cdn.cloudfront.net/image.jpg',
///   transform: TransformOptions(
///     w: 300,
///     h: 200,
///     fit: Fit.cover,
///     format: Format.webp,
///   ),
/// );
/// ```
class SnapkitImageURL {
  final String organizationName;

  SnapkitImageURL(this.organizationName);

  /// Generate Snapkit image proxy URL
  ///
  /// [url] Original image URL
  /// [transform] Image transformation options
  String build({
    required String url,
    TransformOptions? transform,
  }) {
    final uri = Uri.parse('https://$organizationName.snapkit.dev/image');

    final queryParameters = <String, String>{
      'url': url,
    };

    if (transform != null) {
      final transformString = _buildTransformString(transform);
      if (transformString.isNotEmpty) {
        queryParameters['transform'] = transformString;
      }
    }

    return uri.replace(queryParameters: queryParameters).toString();
  }

  String _buildTransformString(TransformOptions options) {
    final parts = <String>[];

    // Numeric/string value parameters
    if (options.w != null) parts.add('w:${options.w}');
    if (options.h != null) parts.add('h:${options.h}');
    if (options.fit != null) parts.add('fit:${options.fit!.name}');
    if (options.format != null) parts.add('format:${options.format!.name}');
    if (options.rotation != null) parts.add('rotation:${options.rotation}');
    if (options.blur != null) parts.add('blur:${options.blur}');
    if (options.dpr != null) parts.add('dpr:${options.dpr}');
    if (options.quality != null) parts.add('quality:${options.quality}');

    // Boolean parameters
    if (options.grayscale == true) parts.add('grayscale');
    if (options.flip == true) parts.add('flip');
    if (options.flop == true) parts.add('flop');

    // Extract parameter
    if (options.extract != null) {
      final e = options.extract!;
      parts.add('extract:${e.x}-${e.y}-${e.width}-${e.height}');
    }

    return parts.join(',');
  }
}

File: swift/Sources/SnapkitImageURL/SnapkitImageURL.swift
import Foundation

/// Image transformation parameters
public struct TransformOptions {
    /// Image width (pixels)
    public var w: Int?
    /// Image height (pixels)
    public var h: Int?
    /// Resize method
    public var fit: Fit?
    /// Output format
    public var format: Format?
    /// Rotation angle (degrees)
    public var rotation: Int?
    /// Blur intensity (0.3-1000)
    public var blur: Int?
    /// Whether to convert to grayscale
    public var grayscale: Bool?
    /// Whether to flip vertically
    public var flip: Bool?
    /// Whether to flip horizontally
    public var flop: Bool?
    /// Region extraction
    public var extract: Extract?
    /// Device Pixel Ratio (1.0-4.0)
    public var dpr: Double?
    /// Image quality (1-100)
    public var quality: Int?

    public init(
        w: Int? = nil,
        h: Int? = nil,
        fit: Fit? = nil,
        format: Format? = nil,
        rotation: Int? = nil,
        blur: Int? = nil,
        grayscale: Bool? = nil,
        flip: Bool? = nil,
        flop: Bool? = nil,
        extract: Extract? = nil,
        dpr: Double? = nil,
        quality: Int? = nil
    ) {
        self.w = w
        self.h = h
        self.fit = fit
        self.format = format
        self.rotation = rotation
        self.blur = blur
        self.grayscale = grayscale
        self.flip = flip
        self.flop = flop
        self.extract = extract
        self.dpr = dpr
        self.quality = quality
    }

    /// Resize method
    public enum Fit: String {
        case contain
        case cover
        case fill
        case inside
        case outside
    }

    /// Output format
    public enum Format: String {
        case jpeg
        case png
        case webp
        case avif
    }

    /// Region extraction
    public struct Extract {
        public let x: Int
        public let y: Int
        public let width: Int
        public let height: Int

        public init(x: Int, y: Int, width: Int, height: Int) {
            self.x = x
            self.y = y
            self.width = width
            self.height = height
        }
    }
}

/// Snapkit image URL builder
public struct SnapkitImageURLBuilder {
    private let organizationName: String

    public init(organizationName: String) {
        self.organizationName = organizationName
    }

    /// Generate Snapkit image proxy URL
    ///
    /// - Parameters:
    ///   - url: Original image URL
    ///   - transform: Image transformation options
    /// - Returns: Complete image proxy URL
    ///
    /// # Example
    /// ```swift
    /// let builder = SnapkitImageURLBuilder(organizationName: "my-org")
    /// let imageURL = builder.build(
    ///     url: "https://cdn.cloudfront.net/image.jpg",
    ///     transform: TransformOptions(
    ///         w: 300,
    ///         h: 200,
    ///         fit: .cover,
    ///         format: .webp
    ///     )
    /// )
    /// ```
    public func build(url: String, transform: TransformOptions? = nil) -> URL? {
        var components = URLComponents(string: "https://\(organizationName).snapkit.dev/image")

        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "url", value: url)
        ]

        if let transform = transform {
            let transformString = buildTransformString(transform)
            if !transformString.isEmpty {
                queryItems.append(URLQueryItem(name: "transform", value: transformString))
            }
        }

        components?.queryItems = queryItems
        return components?.url
    }

    private func buildTransformString(_ options: TransformOptions) -> String {
        var parts: [String] = []

        // Numeric/string value parameters
        if let w = options.w {
            parts.append("w:\(w)")
        }
        if let h = options.h {
            parts.append("h:\(h)")
        }
        if let fit = options.fit {
            parts.append("fit:\(fit.rawValue)")
        }
        if let format = options.format {
            parts.append("format:\(format.rawValue)")
        }
        if let rotation = options.rotation {
            parts.append("rotation:\(rotation)")
        }
        if let blur = options.blur {
            parts.append("blur:\(blur)")
        }
        if let dpr = options.dpr {
            parts.append("dpr:\(dpr)")
        }
        if let quality = options.quality {
            parts.append("quality:\(quality)")
        }

        // Boolean parameters
        if options.grayscale == true {
            parts.append("grayscale")
        }
        if options.flip == true {
            parts.append("flip")
        }
        if options.flop == true {
            parts.append("flop")
        }

        // extract parameter
        if let extract = options.extract {
            parts.append("extract:\(extract.x)-\(extract.y)-\(extract.width)-\(extract.height)")
        }

        return parts.joined(separator: ",")
    }
}
